### ISSUE TITLE
AttributeError: module 'bitsandbytes' has no attribute 'nn'

### ISSUE THREAD

**Issue by WouterMulders1999** [TimDettmers/bitsandbytes#727]
I am working with Amazon Sagemaker with a ml.g5.12x large instance.

In my notebook I install bitsandbytes as follows:

```
!pip uninstall bitsandbytes -y 
!pip install bitsandbytes==0.40.2 #bitsandbytes
%pip install torch
import bitsandbytes
import torch
!python -m bitsandbytes
```

The final line gives me 
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
++++++++++++++++++ BUG REPORT INFORMATION ++++++++++++++++++
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

++++++++++++++++++ /usr/local CUDA PATHS +++++++++++++++++++

+++++++++++++++ WORKING DIRECTORY CUDA PATHS +++++++++++++++

++++++++++++++++++ LD_LIBRARY CUDA PATHS +++++++++++++++++++
 /opt/conda/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/ CUDA PATHS 

++++++++++++++++++++++++++ OTHER +++++++++++++++++++++++++++
COMPILED_WITH_CUDA = True
COMPUTE_CAPABILITIES_PER_GPU = ['8.6', '8.6', '8.6', '8.6']
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
++++++++++++++++++++++ DEBUG INFO END ++++++++++++++++++++++
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Running a quick check that:
    + library is importable
    + CUDA function is callable

WARNING: Please be sure to sanitize sensible info from any such env vars!

SUCCESS!
Installation was successful!

I then try: 
```
from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained(
  'decapoda-research/llama-7b-hf,
  device_map='auto',
  load_in_8bit=True,
  max_memory=f'{int(torch.cuda.mem_get_info()[0]/1024**3)-2}GB')
```

and get `AttributeError: module 'bitsandbytes' has no attribute 'nn'`

This example comes straight from the ReadMe.

I also tried the following approach:

```
model_id = "tiiuae/falcon-40b"
bnb_config = BitsAndBytesConfig(
	load_in_4bit=True,
	bnb_4bit_use_double_quant=True,
	bnb_4bit_quant_type="nf4",
	bnb_4bit_compute_dtype=torch.bfloat16
)
model = AutoModelForCausalLM.from_pretrained(model_id, trust_remote_code=True, quantization_config=bnb_config, device_map="auto")
```

Which results in the exact same `AttributeError: module 'bitsandbytes' has no attribute 'nn'`

Any suggestions on how I can get my code to work? Thanks in advance for your time.

EDIT: I realised below code might be important for this issue too.
```
# Add installed cuda runtime to path for bitsandbytes
cuda_install_dir = '/'.join(nvidia.__file__.split('/')[:-1]) + '/cuda_runtime/lib/'
os.environ['LD_LIBRARY_PATH'] =  cuda_install_dir
```
EDIT 2:
Maybe it goes wrong because of torch.nn?
```
import textract
import textract
import pandas as pd
import random
from random import randrange
from datetime import timedelta
from datetime import datetime
import string
import bitsandbytes
import torch
import torch.nn as nn
from transformers import BitsAndBytesConfig, AutoTokenizer, AutoConfig, AutoModelForCausalLM
import os
import nvidia
```

**Comment by dannstrom** [comment]
I face the same problem. I have fine-tuned the  Llama-2-7b-chat-hf following the instructions given in this notebook: https://colab.research.google.com/drive/1PEQyJO1-f6j0S_XJ8DV50NkpzasXkrzd?usp=sharing#scrollTo=ib_We3NLtj2E

After finished training I reload the Llama-2-7b-chat-hf and merge it with the LORA weights.  Then I save the model to disk. 

When I reload it I use the exact specifications for 'bnb_config' as @WouterMulders1999 and get the error:

  [Previous line repeated 1 more time]
  File "/home/user/miniconda3/envs/env_name/lib/python3.11/site-packages/transformers/integrations/bitsandbytes.py", line 158, in _replace_with_bnb_linear
    model._modules[name] = bnb.nn.Linear4bit(
                           ^^^^^^
AttributeError: module 'bitsandbytes' has no attribute 'nn'

When I "remerge" the models using:
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch.float16,
    device_map=device_map,
)
model = PeftModel.from_pretrained(base_model, new_model)
model = model.merge_and_unload()

Inference works in jupyter notebook but I get the same error in VScode. When trying to load the finished model from disk in jupyter the kernel dies.  

The environment was set up as:
!pip install -q -U bitsandbytes
!pip install -q -U git+https://github.com/huggingface/transformers.git 
!pip install -q -U git+https://github.com/huggingface/peft.git
!pip install -q -U git+https://github.com/huggingface/accelerate.git
!pip install -q datasets
!pip install -q -U einops
!pip install -q -U safetensors
!pip install -q -U torch
!pip install -q -U xformers
!pip install -q -U scipy
!pip install -q -U trl
!pip install -q -U tensorboard
!pip install -q -U tensorboardX

**Comment by TomekPro** [comment]
@TimDettmers the same issue on Google Compute VM ubuntu 20.04. Any idea what could be wrong?

`AttributeError: module 'bitsandbytes' has no attribute 'nn'`

Fails this method from peft:
```
def is_bnb_4bit_available() -> bool:
    if not is_bnb_available():
        return False

    import bitsandbytes as bnb

    return hasattr(bnb.nn, "Linear4bit")
```

But when I do on the poetry venv just `from bitsandbytes import nn` it works.

Libs:
```
peft                      0.6.2
bitsandbytes              0.41.2.post2
torch                     2.1.1
torchvision               0.16.1
transformers              4.35.2
```

Here is the output of python -m bitsandbytes:

```
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
++++++++++++++++++ BUG REPORT INFORMATION ++++++++++++++++++
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

++++++++++++++++++ /usr/local CUDA PATHS +++++++++++++++++++
/usr/local/cuda-12.3/targets/x86_64-linux/lib/stubs/libcuda.so
/usr/local/cuda-12.3/targets/x86_64-linux/lib/libcudart.so

+++++++++++++++ WORKING DIRECTORY CUDA PATHS +++++++++++++++


++++++++++++++++++ LD_LIBRARY CUDA PATHS +++++++++++++++++++
++++++++++ /usr/local/cuda-12.3/lib64 CUDA PATHS +++++++++++


++++++++++++++++++++++++++ OTHER +++++++++++++++++++++++++++
COMPILED_WITH_CUDA = True
COMPUTE_CAPABILITIES_PER_GPU = ['7.5', '7.5']
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
++++++++++++++++++++++ DEBUG INFO END ++++++++++++++++++++++
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Running a quick check that:
    + library is importable
    + CUDA function is callable


WARNING: Please be sure to sanitize sensible info from any such env vars!

SUCCESS!
Installation was successful!
```

@dannstrom @WouterMulders1999 Maybe you've solved it? 

Thanks
Tomek

**Comment by dannstrom** [comment]
@TomekPro unfortunately I haven't. I use ubuntu 20.04 on WSL2 and Jupyter Notebook and it works fine. But I rather use VScode and there it's still a problem. So if you figure it out please let us now :)

**Comment by alirezamborhan** [comment]
Hello,
I also I have this issue with "AttributeError: module 'bitsandbytes' has no attribute 'nn'"

**Comment by matthiasdroth** [comment]
Me too, @alirezamborhan.

**Comment by Rubiel1** [comment]
Hello, I too have this issue. I use Fedora, and I tried to run a LLM on int4 precision.

**Comment by RodolfoPF** [comment]
https://blog.csdn.net/dandandancpop/article/details/135290871
@alirezamborhan, @matthiasdroth, @Rubiel1 
I solved the problem with this tutorial

**Comment by Titus-von-Koeller** [comment]
Hey all, can it be that you have a `bitsandbytes` directory in your current working directory? That would explain it. We should probably migrate to a `src` based layout to avoid such issues.

Please run this script and paste the output here.

```python
import os
import bitsandbytes


# Get the directory of the bitsandbytes module
module_dir = os.path.dirname(bitsandbytes.__file__)

# Print the location of the bitsandbytes module
print(f"bitsandbytes module location: {module_dir}")

# List the contents of the module directory
print(f"Contents of {module_dir}:")
for item in os.listdir(module_dir):
    print(item)
```

**Comment by paullintilhac** [comment]
I have the exact same issue, I installed 0.44 dev from source, and then ran into this when loading lora weights on top of BLIP-2.  @Titus-von-Koeller when I try to run the above code, I get an error on the first line. What action should be taken based on the output? And also, I tried moving my bitsandbytes source code into a separate directory from my working directory in colab, and it didn't have any effect.
